docs/source/custom_sockets.rst

```diff
@@ -0,0 +1,244 @@
+.. _custom-sockets:
+
+=========================
+Custom Sockets in Syncano
+=========================
+
+``Syncano`` gives its users an ability to create custom sockets. It means, that users can define 
+a very specific endpoints in their Syncano application, and use them as other Syncano 
+modules (Classes, Scripts, etc), using standard API calls.
+Currently, custom sockets allow only one dependency - Scripts. It means that on the backend side,
+each API call executes a Script and result of the execution, is returned as a result of the
+API call.
+
+Creating a custom socket
+------------------------
+
+To create a custom socket follow these steps::
+
+    import syncano
+    from syncano.models import CustomSocket, Endpoint, ScriptCall, ScriptDependency, RuntimeChoices
+    from syncano.connection import Connection
+
+    # 1. Initialize a custom socket.
+    custom_socket = CustomSocket(name='my_custom_socket')  # this will create an object in place (do API call)
+
+    # 2. Define endpoints.
+    my_endpoint = Endpoint(name='my_endpoint')  # no API call here
+    my_endpoint.add_call(ScriptCall(name='custom_script'), methods=['GET'])
+    my_endpoint.add_call(ScriptCall(name='another_custom_script'), methods=['POST'])
+
+    # Explanation of the above lines:
+    # Defined endpoint will be visible under `my_endpoint` name:
+    # On this syncano API endpoint the above endpoint will be called (after custom socket creation)
+    # <host>://<api_version>/instances/<instance_name>/endpoints/sockets/my_endpoint/
+    # On this syncano API endpoint the details of the defined endpoint will be returned
+    # <host>://<api_version>/instances/<instance_name>/sockets/my_custom_socket/endpoints/my_endpoint/
+    # For the above endpoint - the two calls are defined, one uses GET method - the custom_script will be executed
+    # there, second uses the POST method and then the another_custom_script will be called;
+    # Currently only script are available for calls;
```

Comment:
Every script consists of dependencies (which are simply checked and created when needed) and endpoints.
Here:
```
    my_endpoint = Endpoint(name='my_endpoint')
    my_endpoint.add_call(ScriptCall(name='custom_script'), methods=['GET'])
    my_endpoint.add_call(ScriptCall(name='another_custom_script'), methods=['POST'])
```
we create one endpoint called my_endpoint and we define that: when someone does GET my_endpoint -> call script with endpoint=custom_script. When someone does POST my_endpoint, call script with endpoint=another_custom_script. So we can define separate logic/bindings based on a HTTP method used.

Now endpoints live in two places in API. 
`/instances/<instance_name>/sockets/my_custom_socket/endpoints/`
Here we will get them as REST objects and we will only get endpoints created by my_custom_socket.
So our endpoint will have URI: `/instances/<instance_name>/sockets/my_custom_socket/endpoints/my_endpoint/`
but this will just be a read_only REST object (with a link to URI that can run it).

Second place defines API for custom sockets in general. 
`/instances/<instance_name>/endpoints/sockets/`
Here you will get all endpoints created by all sockets. This will also be a list of read only REST objects, but if you do an action on a specific endpoint resource URI, e.g.:
`GET /instances/<instance_name>/endpoints/sockets/my_endpoint/`
it will instead be used as a call to that endpoint and as a result - call Script custom_script. While POST will call Script another_custom_script (as these are the only 2 actions that we defined).

It's a little verbose but hopefully I explained it enough?
You could also check: https://github.com/Syncano/syncano-platform/pull/1240 (last 2 sections, endpoints and socket endpoints access)